# Задача на 6-ом шаге урока

**⚖️ Взвешенный блендинг**

Задание: У тебя есть информация о точности модели на лидерборде (accuracy_score). Рассчитай веса, с которыми нужно будет взвесить предсказания моделей, чтобы они были пропорциональны accuracy_score, а их сумма равнялась единице.

Примечание: Ввод и вывод уже написан за вас.

In [3]:
n = int(input())
s = input().split()
scores = []
total = 0

for i in range(n):
    val = float(s[i])
    scores.append(val)
    total += val

weights = []

for i in range(n):
    weight = scores[i] / total
    weights.append(round(weight, 7))

print(*weights)

3
0.3 0.5 0.4
0.25 0.4166667 0.3333333


Sample Input:

3

0.3 0.5 0.4

Sample Output:

0.25 0.4166667 0.3333333

# Задача на 7-ом шаге урока

**🏄‍♂️ Hard voting**

Датасет с прогнозами моделей тут:

In [1]:
path = 'https://stepik.org/media/attachments/lesson/825510/make_soft_blend.csv'

In [2]:
import pandas as pd

In [3]:
df = pd.read_csv(path)

In [9]:
model_cols = [f'model_{i}' for i in range(1, 6)]


df['target_class'] = df[model_cols].mode(axis=1)[0].values

result = df[['car_id', 'target_class']]
result.to_csv('results/submission.csv', index=False)
result

,car_id,target_class
0,car_0,electro_bug
1,car_1,engine_overheat
2,car_2,engine_overheat
3,car_3,electro_bug
4,car_4,engine_check
...,...,...
1908,car_1908,break_bug
1909,car_1909,break_bug
1910,car_1910,engine_overheat
1911,car_1911,engine_check


,model_1,model_2,model_3,model_4,model_5
0,electro_bug,gear_stick,electro_bug,electro_bug,engine_ignition
1,engine_check,engine_overheat,electro_bug,engine_overheat,engine_overheat
2,another_bug,electro_bug,engine_overheat,engine_overheat,gear_stick
3,electro_bug,electro_bug,electro_bug,wheel_shake,electro_bug
4,engine_ignition,another_bug,engine_check,engine_check,engine_check
...,...,...,...,...,...
1908,break_bug,break_bug,engine_overheat,break_bug,another_bug
1909,break_bug,break_bug,wheel_shake,break_bug,gear_stick
1910,engine_overheat,engine_overheat,engine_overheat,engine_overheat,break_bug
1911,engine_check,engine_check,engine_check,electro_bug,break_bug


Файл с решением должен выглядет так с заполненными значениями target_class

In [ ]:
path = 'https://stepik.org/media/attachments/lesson/825510/sb_sample_submission.csv'

In [ ]:
pd.read_csv(path).head()

,car_id,target_class
0,car_0,NaN
1,car_1,NaN
2,car_2,NaN
3,car_3,NaN
4,car_4,NaN


# Задача на 8-ом шаге урока

**🤝Soft Voting**

Датасета с предсказаниями поломок машин от разных моделей:

In [ ]:
'https://stepik.org/media/attachments/lesson/825510/catboost_preds.csv'

In [ ]:
'https://stepik.org/media/attachments/lesson/825510/lgbm_preds.csv'

In [ ]:
'https://stepik.org/media/attachments/lesson/825510/xgb_preds.csv'

Ваша задача  — сблендить предсказания этих трех моделей при помощи методики мягкого голосования (soft voting). В качестве ответа отправьте файл, в котором для каждой строчки будет предсказанный ответ, записанный в колонку soft_preds. Названия ответа можно брать из названия соответствующей колонки, обрезая _probа. Порядок строк в высылаемом решении должен быть таким же, как и в файлах с предсказаниями моделей.



# Задача на 10-ом шаге урока

**Напишите свой Rank Averaging**

Реализуйте процедуру нормализации вероятностей, как описывалось в предыдущем шаге (https://stepik.org/lesson/825510/step/9?unit=829017). Для создания тестового датасета можно воспользоваться кодом:

In [ ]:
import numpy as np
import pandas as pd
np.random.seed(42)
proba = np.random.random(100)
df = pd.DataFrame({'id': np.arange(len(proba)), 'proba': proba})
df

,id,proba
0,0,0.374540
1,1,0.950714
2,2,0.731994
3,3,0.598658
4,4,0.156019
...,...,...
95,95,0.493796
96,96,0.522733
97,97,0.427541
98,98,0.025419


Вам нужно посчитать отнормированные предсказания по инструкции из предыдущего степа. Финальный ответ запишите в столбец normalized_proba

In [ ]:
import numpy as np
import pandas as pd
df['normalized_proba'] = #your code

# Задача на 11-ом шаге урока

**Rank Averaging для 3-х моделей**

В этот раз вам даны вероятности сразу для трех моделей -  proba1, proba2, proba3. Для создания тестового датасета можно использовать код:




In [ ]:
import numpy as np
import pandas as pd
np.random.seed(42)
df = pd.DataFrame({'id': np.arange(100),
                   'proba1': np.random.random(100),
                   'proba2': np.random.random(100),
                   'proba3': np.random.random(100)})

In [ ]:
df

,id,proba1,proba2,proba3
0,0,0.374540,0.031429,0.642032
1,1,0.950714,0.636410,0.084140
2,2,0.731994,0.314356,0.161629
3,3,0.598658,0.508571,0.898554
4,4,0.156019,0.907566,0.606429
...,...,...,...,...
95,95,0.493796,0.349210,0.522243
96,96,0.522733,0.725956,0.769994
97,97,0.427541,0.897110,0.215821
98,98,0.025419,0.887086,0.622890


Посчитайте для каждой модели ее ранг (как в прошлом степе) и потом запишите среднее значение этих трех рангов в колонку mean_rank.

In [ ]:
import numpy as np
import pandas as pd

df['mean_rank'] = #your code